sdks/python/apache_beam/transforms/dofn_lifecycle_test.py

```diff
@@ -0,0 +1,81 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""UnitTests for DoFn lifecycle and bundle methods"""
+
+from __future__ import absolute_import
+
+import unittest
+
+from nose.plugins.attrib import attr
+
+import apache_beam as beam
+from apache_beam.testing.test_pipeline import TestPipeline
+
+class CallSequenceEnforcingDoFn(beam.DoFn):
+  def __init__(self):
+    self._setup_called = False
+    self._start_bundle_calls = 0
+    self._finish_bundle_calls = 0
+    self._teardown_called = False
+
+  def setup(self):
+    assert not self._setup_called,'setup should not be called twice'
+    assert self._start_bundle_calls == 0, 'setup should be called before start_bundle'
+    assert self._finish_bundle_calls == 0, 'setup should be called before finish_bundle'
+    assert not self._teardown_called, 'setup should be called before teardown'
+    self._setup_called = True
+
+  def start_bundle(self):
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls , \
+      'there should be as many start_bundle calls as finish_bundle calls'
+    assert not self._teardown_called, 'teardown should not have been called'
+    self._start_bundle_calls += 1
+
+  def process(self, element)
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls > 0, 'start_bundle should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls + 1, \
+      'there should be one start_bundle call with no call to finish_bundle'
+    assert not self._teardown_called, 'teardown should not have been called'
+    return [element * element]
+
+  def finish_bundle(self):
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls > 0, 'start_bundle should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls + 1, \
+      'there should be one start_bundle call with no call to finish_bundle'
+    assert not self._teardown_called, 'teardown should not have been called'
+    self._finish_bundle_calls += 1
+
+  def teardown(self):
+    assert self._setup_called, 'setup should have been called'
+    assert self._start_bundle_calls == self._finish_bundle_calls , \
+      'there should be as many start_bundle calls as finish_bundle calls'
+    assert not self._teardown_called,'teardown should not be called twice'
+    # Unfortunately there is no way to assert that teardown was actually called.
+    # This is acceptable because teardown is best-effort.
+    self._teardown_called = True
+
+
+@attr('ValidatesRunner')
+class DoFnLifecycleTest(unittest.TestCase):
```

Comment:
Does Python have TestStream implemented in the direct runner? Or something equivalent for testing directly against the SDK harness? It would be good to have a test that explicitly crosses multiple bundles, since the purpose of setup / teardown is for operations that cross bundles.

Reply:
TestStream is implemented in direct runner but it would not use sdk harness.

Also, Yifan's implementation here adds the APIs, but does not really change the lifecycle to re-use dofns.

Nevertheless, it will be good to run a test with sdk harness.

Reply:
Yes, the test already checks that setup is called exactly once, so all that is needed is a way to generate a test such that the runner runs multiple bundles with one DoFn. I'm not sure if that capability exists currently.

Unfortunately the test only checks that teardown is called at most once, and not exactly once. I don't have a good way of checking that.

Reply:
I believe this will be possible if we set _perform_dofn_pickle_test to false in a test. However that needs to be somehow exposed as an option or monkey patched.

Reply:
Added a test that checks that a global flag is set.